grudge/array_context.py

```diff
@@ -163,7 +163,8 @@ class MPIBasedArrayContext:
 # {{{ distributed + pytato
 
 def _to_identifier(s: str) -> str:
-    return "".join(ch for ch in s if ch.isidentifier())
+    # Only allow digits, letters, and underscores in identifiers
+    return "".join(ch for ch in s if ch.isalnum() or ch == "_")
```

Comment:
Huh. I would've expected these two bits of code to be identical. Are they not?

Reply:
They are not identical, a Python identifier can not start with a number (see e.g. https://docs.python.org/3/reference/lexical_analysis.html#identifiers).

```
>>> "1".isidentifier()
False
>>> "1".isalnum()
True
>>> def _to_identifier(s: str) -> str:
...     # Only allow digits, letters, and underscores
...     return "".join(ch for ch in s if ch.isalnum() or ch == "_")
...
>>> _to_identifier("123")
'123'
>>> def _to_identifier_old(s: str) -> str:
...     return "".join(ch for ch in s if ch.isidentifier())
...
>>> _to_identifier_old("123")
''
```

Reply:
(see also https://github.com/inducer/pytools/pull/225 for a different approach)

Reply:
Ah! :facepalm: Makes sense, thanks!

And I think your proposed approach of centralizing this functionality in pytools is reasonable (as you probably guessed from my merging that). I'll roll a pytools release so you can start relying on this.

Reply:
Sounds good! Changed in [3c36e09](https://github.com/inducer/grudge/pull/341/commits/3c36e09b02b84b035ef254b719549975c604237c)